// node-kakao/dist/packet/struct/wrap/channel.js
/*
 * Created on Fri Jan 22 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../../channel", "../../../openlink", "./chat"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.dataStructToOpenChannelInfo = exports.dataStructToNormalChannelInfo = exports.dataStructToChannelInfo = exports.structToOpenChannelInfo = exports.structToNormalChannelInfo = exports.structToChannelInfo = void 0;
    const channel_1 = require("../../../channel");
    const openlink_1 = require("../../../openlink");
    const chat_1 = require("./chat");
    function structToChannelInfo(struct) {
        var _a;
        const displayUserList = struct.displayMembers ? struct.displayMembers.map((userStruct) => {
            return {
                userId: userStruct.userId,
                nickname: userStruct.nickName,
                countryIso: userStruct.countryIso || '',
                profileURL: userStruct.profileImageUrl,
            };
        }) : [];
        const metaMap = {};
        (_a = struct.chatMetas) === null || _a === void 0 ? void 0 : _a.forEach((meta) => metaMap[meta.type] = { ...meta });
        const info = {
            channelId: struct.chatId,
            type: struct.type,
            activeUserCount: struct.activeMembersCount,
            newChatCount: struct.newMessageCount,
            newChatCountInvalid: struct.invalidNewMessageCount,
            lastChatLogId: struct.lastLogId,
            pushAlert: struct.pushAlert,
            lastSeenLogId: struct.lastSeenLogId,
            metaMap,
            displayUserList,
        };
        if (struct.lastChatLog) {
            info['lastChatLog'] = chat_1.structToChatlog(struct.lastChatLog);
        }
        return info;
    }
    exports.structToChannelInfo = structToChannelInfo;
    function structToNormalChannelInfo(struct) {
        return {
            ...structToChannelInfo(struct),
            joinTime: struct.joinedAtForNewMem,
        };
    }
    exports.structToNormalChannelInfo = structToNormalChannelInfo;
    function structToOpenChannelInfo(struct) {
        return {
            ...structToChannelInfo(struct),
            linkId: struct.li,
            openToken: struct.otk,
            directChannel: struct.directChat,
        };
    }
    exports.structToOpenChannelInfo = structToOpenChannelInfo;
    function dataStructToChannelInfo(channelData) {
        const commonInfo = {
            channelId: channelData.c,
            type: channelData.t,
            lastChatLogId: channelData.ll,
            activeUserCount: channelData.a,
            lastSeenLogId: channelData.s,
            newChatCount: channelData.n,
            pushAlert: channelData.p
        };
        if (channelData.l) {
            commonInfo['lastChatLog'] = chat_1.structToChatlog(channelData.l);
        }
        if (channelData.i && channelData.k) {
            commonInfo['displayUserList'];
            const list = [];
            const len = channelData.i.length;
            for (let i = 0; i < len; i++) {
                list.push({ userId: channelData.i[i], nickname: channelData.k[i], profileURL: '' });
            }
        }
        return channel_1.ChannelInfo.createPartial(commonInfo);
    }
    exports.dataStructToChannelInfo = dataStructToChannelInfo;
    function dataStructToNormalChannelInfo(channelData) {
        return channel_1.NormalChannelInfo.createPartial(dataStructToChannelInfo(channelData));
    }
    exports.dataStructToNormalChannelInfo = dataStructToNormalChannelInfo;
    function dataStructToOpenChannelInfo(channelData) {
        return openlink_1.OpenChannelInfo.createPartial({
            ...dataStructToChannelInfo(channelData),
            linkId: channelData.li,
            openToken: channelData.otk
        });
    }
    exports.dataStructToOpenChannelInfo = dataStructToOpenChannelInfo;
});
