// node-kakao/dist/packet/struct/wrap/openlink.js
/*
 * Created on Mon Jan 25 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./user"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.structToOpenLinkInfo = exports.structToOpenLink = void 0;
    const user_1 = require("./user");
    function structToOpenLink(struct) {
        return {
            linkId: struct.li,
            linkName: struct.ln,
            searchable: struct.sc,
            type: struct.lt,
            openToken: struct.otk,
            linkCoverURL: struct.lcu,
            linkURL: struct.lu,
            createdAt: struct.ca,
            description: struct.desc,
            activated: struct.ac,
            privilege: struct.pv,
            profileTagList: struct.op && struct.op.tags || [],
            linkOwner: user_1.structToOpenLinkChannelUserInfo(struct.olu),
        };
    }
    exports.structToOpenLink = structToOpenLink;
    function structToOpenLinkInfo(struct) {
        return {
            directLimit: struct.dcl || 0,
            userLimit: struct.ml || 0,
        };
    }
    exports.structToOpenLinkInfo = structToOpenLinkInfo;
});
