// node-kakao/dist/talk/network/index.js
/*
 * Created on Fri Jan 22 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../crypto", "../../network", "../../request", "../../network/socket"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkSessionFactory = void 0;
    const crypto_1 = require("../../crypto");
    const network_1 = require("../../network");
    const request_1 = require("../../request");
    const NetSocket = __importStar(require("../../network/socket"));
    /**
     * Create loco stream by performing booking and checkin.
     */
    class TalkSessionFactory {
        async getConf(config) {
            const bookingStream = await NetSocket.createTLSSocket({
                host: config.locoBookingHost,
                port: config.locoBookingPort,
                keepAlive: false,
            });
            return network_1.getBookingData(bookingStream, config);
        }
        async getCheckin(userId, config) {
            let checkinStream;
            const checkinCrypto = await crypto_1.newCryptoStore(config.locoPEMPublicKey);
            try {
                const conf = await this.getConf(config);
                if (!conf.success)
                    return conf;
                checkinStream = new network_1.LocoSecureLayer(await NetSocket.createTCPSocket({
                    host: conf.result.ticket.lsl[0],
                    port: conf.result.wifi.ports[0],
                    keepAlive: false,
                }), checkinCrypto);
            }
            catch (e) {
                if (!config.locoCheckinFallbackHost || !config.locoCheckinFallbackPort)
                    throw e;
                // Fallback
                checkinStream = new network_1.LocoSecureLayer(await NetSocket.createTCPSocket({
                    host: config.locoCheckinFallbackHost,
                    port: config.locoCheckinFallbackPort,
                    keepAlive: false,
                }), checkinCrypto);
            }
            return network_1.getCheckinData(checkinStream, config, userId);
        }
        async connect(userId, config) {
            const checkinRes = await this.getCheckin(userId, config);
            if (!checkinRes.success)
                return checkinRes;
            const locoStream = new network_1.LocoSecureLayer(await NetSocket.createTCPSocket({
                host: checkinRes.result.host,
                port: checkinRes.result.port,
                keepAlive: true,
            }), await crypto_1.newCryptoStore(config.locoPEMPublicKey));
            return { status: request_1.KnownDataStatusCode.SUCCESS, success: true, result: new network_1.LocoSession(locoStream) };
        }
    }
    exports.TalkSessionFactory = TalkSessionFactory;
});
