// node-kakao/dist/api/attachment-api-client.js
/*
 * Created on Thu Feb 18 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../chat", "../config", "../request", "./header-util", "./web-client"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AttachmentApi = exports.AttachmentApiClient = void 0;
    const chat_1 = require("../chat");
    const config_1 = require("../config");
    const request_1 = require("../request");
    const header_util_1 = require("./header-util");
    const web_client_1 = require("./web-client");
    /**
     * Web attachment upload api
     */
    class AttachmentApiClient {
        constructor(mediaClient, videoClient, audioClient, config) {
            this.config = config;
            this._mediaClient = new web_client_1.TextWebRequest(mediaClient);
            this._videoClient = new web_client_1.TextWebRequest(videoClient);
            this._audioClient = new web_client_1.TextWebRequest(audioClient);
        }
        createHeader() {
            const header = {};
            header_util_1.fillBaseHeader(header, this.config);
            const userAgent = header_util_1.getUserAgent(this.config);
            header['User-Agent'] = userAgent;
            return header;
        }
        async upload(type, filename, data) {
            const client = this.getReqClient(type);
            const mimeType = this.getMimeType(type);
            const res = await client.requestMultipartText('POST', 'upload', {
                'user_id': 0,
                'attachment_type': mimeType,
                'attachment': {
                    value: data,
                    options: {
                        filename
                    }
                }
            }, this.createHeader());
            return {
                success: true,
                status: request_1.KnownDataStatusCode.SUCCESS,
                result: {
                    path: res,
                    s: data.byteLength
                }
            };
        }
        getReqClient(type) {
            switch (type) {
                case chat_1.KnownChatType.VIDEO: return this._videoClient;
                case chat_1.KnownChatType.AUDIO: return this._audioClient;
                default: return this._mediaClient;
            }
        }
        getMimeType(type) {
            switch (type) {
                case chat_1.KnownChatType.PHOTO: return 'image/jpeg';
                case chat_1.KnownChatType.MULTIPHOTO: return 'image/jpeg';
                case chat_1.KnownChatType.CONTACT: return 'text/x-vcard';
                case chat_1.KnownChatType.VIDEO: return 'video/mp4';
                case chat_1.KnownChatType.AUDIO: return 'audio/m4a';
                // application/octet-stream
                default: return 'image/jpeg';
            }
        }
        /**
         * Create default AttachmentApiClient using config.
         *
         * @param {Partial<WebApiConfig>} config
         */
        static async create(config = {}) {
            return new AttachmentApiClient(await web_client_1.createWebClient('https', 'up-m.talk.kakao.com'), await web_client_1.createWebClient('https', 'up-v.talk.kakao.com'), await web_client_1.createWebClient('https', 'up-a.talk.kakao.com'), Object.assign({ ...config_1.DefaultConfiguration }, config));
        }
    }
    exports.AttachmentApiClient = AttachmentApiClient;
    var AttachmentApi;
    (function (AttachmentApi) {
        let client = null;
        async function upload(type, filename, data) {
            if (!client)
                client = await AttachmentApiClient.create();
            return client.upload(type, filename, data);
        }
        AttachmentApi.upload = upload;
    })(AttachmentApi = exports.AttachmentApi || (exports.AttachmentApi = {}));
});
