// node-kakao/dist/api/auth-api-client.js
/*
 * Created on Thu Jan 28 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./web-client", "../config", "../request", "./header-util", "./struct", "./xvc"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AuthApiClient = exports.KnownAuthStatusCode = void 0;
    const web_client_1 = require("./web-client");
    const config_1 = require("../config");
    const request_1 = require("../request");
    const header_util_1 = require("./header-util");
    const struct_1 = require("./struct");
    const xvc_1 = require("./xvc");
    /**
     * Status code for auth client results
     */
    var KnownAuthStatusCode;
    (function (KnownAuthStatusCode) {
        KnownAuthStatusCode[KnownAuthStatusCode["LOGIN_FAILED_REASON"] = 12] = "LOGIN_FAILED_REASON";
        KnownAuthStatusCode[KnownAuthStatusCode["TOO_MANY_TRY_LOGIN"] = 13] = "TOO_MANY_TRY_LOGIN";
        KnownAuthStatusCode[KnownAuthStatusCode["LOGIN_FAILED"] = 30] = "LOGIN_FAILED";
        KnownAuthStatusCode[KnownAuthStatusCode["MOBILE_UNREGISTERED"] = 32] = "MOBILE_UNREGISTERED";
        KnownAuthStatusCode[KnownAuthStatusCode["DEVICE_NOT_REGISTERED"] = -100] = "DEVICE_NOT_REGISTERED";
        KnownAuthStatusCode[KnownAuthStatusCode["ANOTHER_LOGON"] = -101] = "ANOTHER_LOGON";
        KnownAuthStatusCode[KnownAuthStatusCode["DEVICE_REGISTER_FAILED"] = -102] = "DEVICE_REGISTER_FAILED";
        KnownAuthStatusCode[KnownAuthStatusCode["INVALID_DEVICE_REGISTER"] = -110] = "INVALID_DEVICE_REGISTER";
        KnownAuthStatusCode[KnownAuthStatusCode["INCORRECT_PASSCODE"] = -111] = "INCORRECT_PASSCODE";
        KnownAuthStatusCode[KnownAuthStatusCode["PASSCODE_REQUEST_FAILED"] = -112] = "PASSCODE_REQUEST_FAILED";
        KnownAuthStatusCode[KnownAuthStatusCode["ACCOUNT_RESTRICTED"] = -997] = "ACCOUNT_RESTRICTED";
    })(KnownAuthStatusCode = exports.KnownAuthStatusCode || (exports.KnownAuthStatusCode = {}));
    /**
     * Provides default pc login api which can obtain OAuthCredential
     */
    class AuthApiClient {
        constructor(client, _name, _deviceUUID, config, xvcProvider) {
            this._name = _name;
            this._deviceUUID = _deviceUUID;
            this.config = config;
            this.xvcProvider = xvcProvider;
            this._client = new web_client_1.DataWebRequest(client);
        }
        get name() {
            return this._name;
        }
        get deviceUUID() {
            return this._deviceUUID;
        }
        async createAuthHeader(form) {
            const header = {};
            header_util_1.fillBaseHeader(header, this.config);
            header_util_1.fillAHeader(header, this.config);
            const userAgent = header_util_1.getUserAgent(this.config);
            header['User-Agent'] = userAgent;
            header['X-VC'] = await this.calculateXVCKey(this.deviceUUID, userAgent, form.email);
            return header;
        }
        fillAuthForm(form) {
            form['device_uuid'] = this._deviceUUID;
            form['device_name'] = this._name;
            if (this.config.deviceModel) {
                form['model_name'] = this.config.deviceModel;
            }
            return form;
        }
        /**
         * Login using given data.
         *
         * @param {LoginForm} form
         */
        async login(form) {
            const res = await this._client.requestData('POST', this.getApiPath('login.json'), this.fillAuthForm({ ...form }), await this.createAuthHeader(form));
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { status: res.status, success: false };
            return {
                status: res.status,
                success: true,
                result: struct_1.structToLoginData(res, this._deviceUUID),
            };
        }
        /**
         * Login using token.
         *
         * @param {TokenLoginForm} form
         */
        async loginToken(form) {
            const res = await this._client.requestData('POST', this.getApiPath('login.json'), this.fillAuthForm({ ...form, auto_login: true }), await this.createAuthHeader(form));
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { status: res.status, success: false };
            return {
                status: res.status,
                success: true,
                result: struct_1.structToLoginData(res, this._deviceUUID),
            };
        }
        /**
         * Request passcode
         *
         * @param {LoginForm} form
         */
        async requestPasscode(form) {
            const res = await this._client.requestData('POST', this.getApiPath('request_passcode.json'), this.fillAuthForm({ ...form }), await this.createAuthHeader(form));
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        /**
         * Try to register device with passcode
         *
         * @param {LoginForm} form
         * @param {string} passcode
         * @param {boolean} [permanent=true] If true the device will be registered as permanent
         */
        async registerDevice(form, passcode, permanent = true) {
            const res = await this._client.requestData('POST', this.getApiPath('register_device.json'), this.fillAuthForm({ ...form, passcode, permanent }), await this.createAuthHeader(form));
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async calculateXVCKey(deviceUUID, userAgent, email) {
            return (await this.xvcProvider.toFullXVCKey(deviceUUID, userAgent, email)).substring(0, 16);
        }
        getApiPath(api) {
            return `${this.config.agent}/account/${api}`;
        }
        /**
         * Create default AuthClient using config.
         *
         * @param {string} name
         * @param {string} deviceUUID
         * @param {Partial<OAuthLoginConfig>} config
         * @param {XVCProvider} [xvcProvider]
         */
        static async create(name, deviceUUID, config = {}, xvcProvider) {
            return new AuthApiClient(await web_client_1.createWebClient('https', 'katalk.kakao.com'), name, deviceUUID, Object.assign({ ...config_1.DefaultConfiguration }, config), xvcProvider ? xvcProvider : xvc_1.Win32XVCProvider);
        }
    }
    exports.AuthApiClient = AuthApiClient;
});
