// node-kakao/dist/api/axios-web-client.js
/*
 * Created on Thu Jan 28 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "axios", "./web-api-util", "form-data"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AxiosWebClient = void 0;
    const axios_1 = __importDefault(require("axios"));
    const web_api_util_1 = require("./web-api-util");
    const form_data_1 = __importDefault(require("form-data"));
    /**
     * WebClient implementation wrapped with axios
     */
    class AxiosWebClient {
        constructor(scheme, host, _decorator) {
            this.scheme = scheme;
            this.host = host;
            this._decorator = _decorator;
        }
        get url() {
            return `${this.scheme}://${this.host}`;
        }
        toApiURL(path) {
            return `${this.url}/${path}`;
        }
        fillHeader(header) {
            var _a;
            header['Host'] = this.host;
            (_a = this._decorator) === null || _a === void 0 ? void 0 : _a.fillHeader(header);
        }
        buildAxiosReqData(method, header) {
            const headers = {};
            this.fillHeader(headers);
            const reqData = {
                headers,
                method,
                // https://github.com/axios/axios/issues/811
                // https://github.com/axios/axios/issues/907
                transformResponse: (data) => data,
                responseType: 'arraybuffer',
                maxContentLength: 100000000,
                maxBodyLength: 100000000
            };
            if (header)
                Object.assign(headers, header);
            return reqData;
        }
        async request(method, path, form, headers) {
            const reqData = this.buildAxiosReqData(method, headers);
            reqData.url = this.toApiURL(path);
            if (form) {
                const formData = web_api_util_1.convertToFormData(form);
                reqData.data = formData.toString();
            }
            const res = await axios_1.default.request(reqData);
            if (res.status !== 200) {
                throw new Error(`Web request failed with status ${res.status} ${res.statusText}`);
            }
            return res.data;
        }
        async requestMultipart(method, path, form, headers) {
            const reqData = this.buildAxiosReqData(method, headers);
            reqData.url = this.toApiURL(path);
            if (form) {
                const formData = this.convertToMultipart(form);
                Object.assign(reqData.headers, formData.getHeaders());
                reqData.data = formData.getBuffer();
            }
            const res = await axios_1.default.request(reqData);
            if (res.status !== 200) {
                throw new Error(`Web request failed with status ${res.status} ${res.statusText}`);
            }
            return res.data;
        }
        convertToMultipart(form) {
            const formData = new form_data_1.default();
            for (const [key, value] of Object.entries(form)) {
                if (value && value.value && value.options) {
                    const file = value;
                    formData.append(key, Buffer.from(file.value), file.options);
                }
                else {
                    formData.append(key, value + '');
                }
            }
            return formData;
        }
    }
    exports.AxiosWebClient = AxiosWebClient;
});
