// node-kakao/dist/api/oauth-api-client.js
/*
 * Created on Sat Feb 13 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../config", "../request", "./header-util", "./web-client"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.OAuthApiClient = void 0;
    const config_1 = require("../config");
    const request_1 = require("../request");
    const header_util_1 = require("./header-util");
    const web_client_1 = require("./web-client");
    class OAuthApiClient {
        constructor(client, config) {
            this.config = config;
            this._client = new web_client_1.DataWebRequest(client);
        }
        createOAuthHeader(credential) {
            const header = {};
            header_util_1.fillBaseHeader(header, this.config);
            header_util_1.fillAHeader(header, this.config);
            header_util_1.fillCredential(header, credential);
            const userAgent = header_util_1.getUserAgent(this.config);
            header['User-Agent'] = userAgent;
            return header;
        }
        /**
         * Renew oauth credential using refresh token
         *
         * @param {OAuthCredential} credential
         */
        async renew(credential) {
            const res = await this._client.requestData('POST', `${this.config.agent}/account/oauth2_token.json`, {
                'grant_type': 'refresh_token',
                'access_token': credential.accessToken,
                'refresh_token': credential.refreshToken
            }, this.createOAuthHeader(credential));
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS) {
                return { status: res.status, success: false };
            }
            return {
                status: res.status,
                success: true,
                result: {
                    type: res['token_type'],
                    credential: {
                        userId: credential.userId,
                        deviceUUID: credential.deviceUUID,
                        accessToken: res['access_token'],
                        refreshToken: res['refresh_token']
                    },
                    expiresIn: res['expires_in']
                }
            };
        }
        /**
         * Create default OAuthApiClient using credential and config.
         *
         * @param {Partial<WebApiConfig>} config
         */
        static async create(config = {}) {
            return new OAuthApiClient(await web_client_1.createWebClient('https', 'katalk.kakao.com'), Object.assign({ ...config_1.DefaultConfiguration }, config));
        }
    }
    exports.OAuthApiClient = OAuthApiClient;
});
