// node-kakao/dist/api/open-upload-api-client.js
/*
 * Created on Sat Apr 10 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../request", "./web-client"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.OpenUploadAPI = exports.OpenUploadApiClient = void 0;
    const request_1 = require("../request");
    const web_client_1 = require("./web-client");
    /**
     * OpenLink profile, link post upload api
     */
    class OpenUploadApiClient {
        constructor(client) {
            this._client = new web_client_1.TextWebRequest(client);
        }
        /**
         * Upload openlink profile image
         *
         * @param {string} name filename
         * @param {Uint8Array} file
         * @param {string} [contentType='image/jpeg']
         * @return {AsyncCommandResult<OpenProfileUploadStruct>}
         */
        async uploadLinkImage(name, file, contentType = 'image/jpeg') {
            const res = await this._client.requestText('POST', 'up/open-chat-profile', { 'file_1': { value: file, options: { filename: name, contentType: contentType } } });
            try {
                const json = JSON.parse(res);
                return { status: request_1.KnownDataStatusCode.SUCCESS, success: true, result: json };
            }
            catch (e) {
                return { status: request_1.KnownDataStatusCode.OPERATION_DENIED, success: false };
            }
        }
        /**
         * Upload openlink profile post image
         *
         * @param {string} name filename
         * @param {Uint8Array} file
         * @param {string} [contentType='image/jpeg']
         * @return {AsyncCommandResult<OpenProfileUploadStruct>}
         */
        async uploadLinkPostImage(name, file, contentType = 'image/jpeg') {
            const res = await this._client.requestText('POST', 'up/open-chat-profile-post', { 'file_1': { value: file, options: { filename: name, contentType: contentType } } });
            try {
                const json = JSON.parse(res);
                return { status: request_1.KnownDataStatusCode.SUCCESS, success: true, result: json };
            }
            catch (e) {
                return { status: request_1.KnownDataStatusCode.OPERATION_DENIED, success: false };
            }
        }
        /**
         * Create default OpenUploadApiClient.
         */
        static async create() {
            return new OpenUploadApiClient(await web_client_1.createWebClient('https', 'up-api1-kage.kakao.com'));
        }
    }
    exports.OpenUploadApiClient = OpenUploadApiClient;
    var OpenUploadAPI;
    (function (OpenUploadAPI) {
        let client = null;
        function getOriginalLinkImageURL(accessKey) {
            return `http://open.kakaocdn.net/dn/${accessKey}/img.jpg`;
        }
        OpenUploadAPI.getOriginalLinkImageURL = getOriginalLinkImageURL;
        function getSmallLinkImageURL(accessKey) {
            return `http://open.kakaocdn.net/dn/${accessKey}/img_s.jpg`;
        }
        OpenUploadAPI.getSmallLinkImageURL = getSmallLinkImageURL;
        function getLargeLinkImageURL(accessKey) {
            return `http://open.kakaocdn.net/dn/${accessKey}/img_l.jpg`;
        }
        OpenUploadAPI.getLargeLinkImageURL = getLargeLinkImageURL;
        /**
         * Upload profile image and construct OpenLink anon profile using nickname.
         *
         * @param {string} nickname Profile nickname
         * @param {Uint8Array} profile Profile image
         * @return {AsyncCommandResult<OpenLinkAnonProfile>}
         */
        async function buildProfile(nickname, profile) {
            if (!client)
                client = await OpenUploadApiClient.create();
            const res = await client.uploadLinkImage('profile.png', profile);
            if (!res.success)
                return res;
            return {
                status: request_1.KnownDataStatusCode.SUCCESS,
                success: true,
                result: { nickname, profilePath: res.result.access_key }
            };
        }
        OpenUploadAPI.buildProfile = buildProfile;
    })(OpenUploadAPI = exports.OpenUploadAPI || (exports.OpenUploadAPI = {}));
});
