// node-kakao/dist/api/service-api-client.js
/*
 * Created on Sun Jan 31 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "bson", "./web-client", "../config", "../request", "../util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ServiceApiClient = void 0;
    const bson_1 = require("bson");
    const web_client_1 = require("./web-client");
    const config_1 = require("../config");
    const request_1 = require("../request");
    const util_1 = require("../util");
    class ServiceApiClient {
        constructor(client) {
            this._client = new web_client_1.DataWebRequest(client);
        }
        get config() {
            return this._client.client.config;
        }
        set config(config) {
            this._client.client.config = config;
        }
        // account
        /**
         * Request more settings. Official client sends this after login
         *
         * @param {any} since Unknown
         */
        async requestMoreSettings(since = 0) {
            const res = await this._client.requestData('GET', 
            // eslint-disable-next-line max-len
            `${this.getAccountApiPath('more_settings.json')}?since=${encodeURIComponent(since)}&lang=${encodeURIComponent(this.config.language)}`);
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        /**
         * Request simplified settings. Official client sends this after login
         *
         * @param {any} since Unknown
         */
        async requestLessSettings(since = 0) {
            const res = await this._client.requestData('GET', 
            // eslint-disable-next-line max-len
            `${this.getAccountApiPath('less_settings.json')}?since=${encodeURIComponent(since)}&lang=${encodeURIComponent(this.config.language)}`);
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async updateSettings(settings) {
            const res = await this._client.requestData('POST', this.getAccountApiPath('update_settings.json'), settings);
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        /**
         * Get one time web login token.
         *
         * Use @method requestSessionURL to get complete url.
         */
        async requestWebLoginToken() {
            const res = await this._client.requestData('GET', this.getAccountApiPath('login_token.json'));
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        /**
         * Create session url. Redirect to redirectURL with session info included.
         *
         * @param {string} redirectURL
         */
        async requestSessionURL(redirectURL) {
            const res = await this.requestWebLoginToken();
            if (!res.success)
                return res;
            return {
                status: res.status,
                success: true,
                result: ServiceApiClient.createSessionURL(res.result.token, redirectURL),
            };
        }
        async canChangeUUID(uuid) {
            const res = await this._client.requestData('POST', this.getAccountApiPath('can_change_uuid.json'), { uuid: uuid });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS, result: res };
        }
        async changeUUID(uuid) {
            const res = await this._client.requestData('POST', this.getAccountApiPath('change_uuid.json'), { uuid: uuid });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS, result: res };
        }
        // friends
        async addFriend(id, pa = '') {
            const res = await this._client.requestData('GET', `${this.getFriendsApiPath('add')}/${encodeURIComponent(id.toString())}.json?pa=${encodeURIComponent(pa)}`);
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async addFriendWithPhoneNumber(nickname, countryIso, countryCode, phoneNumber) {
            const res = await this._client.requestData('POST', this.getFriendsApiPath('add_by_phonenumber.json'), {
                nickname: nickname,
                country_iso: countryIso,
                country_code: countryCode,
                phonenumber: phoneNumber,
            });
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async removeFriend(id) {
            const res = await this._client.requestData('POST', this.getFriendsApiPath('purge.json'), { id: id.toString() });
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async removeFriendList(idList) {
            const res = await this._client.requestData('POST', this.getFriendsApiPath('delete.json'), { ids: util_1.JsonUtil.stringifyLoseless(idList) });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async hideFriend(id, pa = '') {
            const res = await this._client.requestData('POST', this.getFriendsApiPath('hide.json'), { id: id.toString(), pa: pa });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async unhideFriend(id) {
            const res = await this._client.requestData('POST', this.getFriendsApiPath('unhide.json'), { id: id.toString() });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async searchFriends(query, pageNum, pageSize) {
            let res;
            if (pageNum && pageSize) {
                res = await this._client.requestData('GET', this.getFriendsApiPath('search.json'), { query: query, page_num: pageNum, page_size: pageSize });
            }
            else {
                res = await this._client.requestData('GET', this.getFriendsApiPath('search.json'), { query });
            }
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async findFriendById(id) {
            const res = await this._client.requestData('GET', this.getFriendsApiPath(`${id.toString()}.json`));
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async findFriendByUUID(uuid) {
            const res = await this._client.requestData('POST', `${this.getFriendsApiPath('find_by_uuid.json')}`, { uuid: uuid });
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async requestFriendList(types = ['plus', 'normal'], eventTypes = ['create'], token = bson_1.Long.ZERO) {
            const res = await this._client.requestData('GET', `${this.getFriendsApiPath('list.json')}`, { type: JSON.stringify(types), event_types: JSON.stringify(eventTypes), token });
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async setNickname(id, nickname) {
            const res = await this._client.requestData('POST', this.getFriendsApiPath('nickname.json'), { id: id.toString(), nickname: nickname });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async addFavoriteFriends(idList) {
            const res = await this._client.requestData('POST', this.getFriendsApiPath('add_favorite.json'), { ids: util_1.JsonUtil.stringifyLoseless(idList) });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async removeFavoriteFriend(id) {
            const res = await this._client.requestData('POST', this.getFriendsApiPath('remove_favorite.json'), { id: id.toString() });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        // profile
        async requestMusicList(id) {
            const res = await this._client.requestData('GET', this.getProfileApiPath('music/list.json'), { id: id.toString() });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS, result: res };
        }
        async requestMyProfile() {
            const res = await this._client.requestData('GET', this.getProfile3ApiPath('me.json'));
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res,
            };
        }
        async requestProfile(id) {
            const res = await this._client.requestData('GET', `${this.getProfile3ApiPath('friend_info.json')}?id=${encodeURIComponent(id.toString())}`);
            return {
                status: res.status,
                success: res.status === request_1.KnownDataStatusCode.SUCCESS,
                result: res
            };
        }
        // scrap
        async getPreviewURL(url) {
            const res = await this._client.requestData('POST', this.getScrapApiPath('preview.json'), { url });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS, result: res };
        }
        getAccountApiPath(api) {
            return `${this.config.agent}/account/${api}`;
        }
        getFriendsApiPath(api) {
            return `${this.config.agent}/friends/${api}`;
        }
        getProfileApiPath(api) {
            return `${this.config.agent}/profile/${api}`;
        }
        getProfile3ApiPath(api) {
            return `${this.config.agent}/profile3/${api}`;
        }
        getScrapApiPath(api) {
            return `${this.config.agent}/scrap/${api}`;
        }
        /**
         * Create default AccountClient using credential and config.
         *
         * @param {OAuthCredential} credential
         * @param {Partial<WebApiConfig>} config
         */
        static async create(credential, config = {}) {
            return new ServiceApiClient(await web_client_1.createSessionWebClient(credential, Object.assign({ ...config_1.DefaultConfiguration }, config), 'https', 'katalk.kakao.com'));
        }
        static createSessionURL(token, redirectURL) {
            // eslint-disable-next-line max-len
            return `https://accounts.kakao.com/weblogin/login_redirect?continue=${encodeURIComponent(redirectURL)}&token=${token}`;
        }
    }
    exports.ServiceApiClient = ServiceApiClient;
});
