// node-kakao/dist/api/web-client.js
/*
 * Created on Mon Feb 01 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../util", "../util/platform", "./header-util"], factory);
    }
})(function (require, exports) {
    "use strict";
    var __syncRequire = typeof module === "object" && typeof module.exports === "object";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.createSessionWebClient = exports.createWebClient = exports.SessionWebClient = exports.DataWebRequest = exports.TextWebRequest = void 0;
    const util_1 = require("../util");
    const platform_1 = require("../util/platform");
    const header_util_1 = require("./header-util");
    class TextWebRequest {
        constructor(_client) {
            this._client = _client;
        }
        get client() {
            return this._client;
        }
        async requestText(method, path, form, headers) {
            const res = await this._client.request(method, path, form, headers);
            return new TextDecoder('utf-8').decode(res);
        }
        async requestMultipartText(method, path, form, headers) {
            const res = await this._client.requestMultipart(method, path, form, headers);
            return new TextDecoder('utf-8').decode(res);
        }
    }
    exports.TextWebRequest = TextWebRequest;
    class DataWebRequest {
        constructor(client) {
            this._client = new TextWebRequest(client);
        }
        get client() {
            return this._client.client;
        }
        async requestData(method, path, form, headers) {
            const res = await this._client.requestText(method, path, form, headers);
            return util_1.JsonUtil.parseLoseless(res);
        }
        async requestMultipartData(method, path, form, headers) {
            const res = await this._client.requestMultipartText(method, path, form, headers);
            return util_1.JsonUtil.parseLoseless(res);
        }
    }
    exports.DataWebRequest = DataWebRequest;
    /**
     * Api client with credential
     */
    class SessionWebClient {
        constructor(_client, _credential, config) {
            this._client = _client;
            this._credential = _credential;
            this.config = config;
        }
        fillHeader(header) {
            header_util_1.fillCredential(header, this._credential);
        }
        get url() {
            return this._client.url;
        }
        createSessionHeader(headers) {
            const credentialHeader = headers ? { ...headers } : {};
            this.fillHeader(credentialHeader);
            header_util_1.fillBaseHeader(credentialHeader, this.config);
            header_util_1.fillAHeader(credentialHeader, this.config);
            const userAgent = header_util_1.getUserAgent(this.config);
            credentialHeader['User-Agent'] = userAgent;
            return credentialHeader;
        }
        request(method, path, form, headers) {
            return this._client.request(method, path, form, this.createSessionHeader(headers));
        }
        requestMultipart(method, path, form, headers) {
            return this._client.requestMultipart(method, path, form, this.createSessionHeader(headers));
        }
    }
    exports.SessionWebClient = SessionWebClient;
    /**
     * Create web client by platform
     *
     * @param {string} scheme
     * @param {string} host
     * @param {HeaderDecorator} decorator
     */
    async function createWebClient(scheme, host, decorator) {
        if (platform_1.isNode()) {
            return new (await (__syncRequire ? Promise.resolve().then(() => __importStar(require('./axios-web-client'))) : new Promise((resolve_1, reject_1) => { require(['./axios-web-client'], resolve_1, reject_1); }).then(__importStar))).AxiosWebClient(scheme, host, decorator);
        }
        else if (platform_1.isDeno()) {
            return new (await (__syncRequire ? Promise.resolve().then(() => __importStar(require('./fetch-web-client'))) : new Promise((resolve_2, reject_2) => { require(['./fetch-web-client'], resolve_2, reject_2); }).then(__importStar))).FetchWebClient(scheme, host, decorator);
        }
        else if (platform_1.isBrowser()) {
            return new (await (__syncRequire ? Promise.resolve().then(() => __importStar(require('./fetch-web-client'))) : new Promise((resolve_3, reject_3) => { require(['./fetch-web-client'], resolve_3, reject_3); }).then(__importStar))).FetchWebClient(scheme, host, decorator);
        }
        else {
            throw new Error('Unknown environment');
        }
    }
    exports.createWebClient = createWebClient;
    async function createSessionWebClient(credential, config, scheme, host, decorator) {
        return new SessionWebClient(await createWebClient(scheme, host, decorator), credential, config);
    }
    exports.createSessionWebClient = createSessionWebClient;
});
