// node-kakao/dist/talk/openlink/talk-open-channel.js
/*
 * Created on Wed Jan 27 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../event", "../../request", "../channel", "./talk-open-channel-session", "./talk-open-channel-handler", "../../util", "../../channel/meta", "./talk-open-channel-data-session"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkOpenChannel = void 0;
    const event_1 = require("../../event");
    const request_1 = require("../../request");
    const channel_1 = require("../channel");
    const talk_open_channel_session_1 = require("./talk-open-channel-session");
    const talk_open_channel_handler_1 = require("./talk-open-channel-handler");
    const util_1 = require("../../util");
    const meta_1 = require("../../channel/meta");
    const talk_open_channel_data_session_1 = require("./talk-open-channel-data-session");
    class TalkOpenChannel extends event_1.TypedEmitter {
        constructor(_channel, session, store, _chatListStore) {
            super();
            this._channel = _channel;
            this._chatListStore = _chatListStore;
            this._channelSession = new channel_1.TalkChannelDataSession(session.clientUser, new channel_1.TalkChannelSession(this, session), store, _chatListStore);
            const openSession = new talk_open_channel_session_1.TalkOpenChannelSession(this, session);
            this._openChannelSession = new talk_open_channel_data_session_1.TalkOpenChannelDataSession(session.clientUser, openSession, store);
            this._handler = new channel_1.TalkChannelHandler(this, this, store, _chatListStore);
            this._openHandler = new talk_open_channel_handler_1.TalkOpenChannelHandler(this, openSession, this, store, _chatListStore);
        }
        get clientUser() {
            return this._openChannelSession.clientUser;
        }
        get channelId() {
            return this._channel.channelId;
        }
        get chatListStore() {
            return this._chatListStore;
        }
        get store() {
            return this._openChannelSession.store;
        }
        get linkId() {
            return this.store.info.linkId;
        }
        get info() {
            return this.store.info;
        }
        get userCount() {
            return this.store.userCount;
        }
        getName() {
            const nameMeta = this.info.metaMap[meta_1.KnownChannelMetaType.TITLE];
            return nameMeta && nameMeta.content || '';
        }
        getDisplayName() {
            var _a;
            return this.getName() || ((_a = this.info.openLink) === null || _a === void 0 ? void 0 : _a.linkName) || '';
        }
        getUserInfo(user) {
            return this.store.getUserInfo(user);
        }
        getAllUserInfo() {
            return this.store.getAllUserInfo();
        }
        getReadCount(chat) {
            return this.store.getReadCount(chat);
        }
        getReaders(chat) {
            return this.store.getReaders(chat);
        }
        async sendChat(chat, noSeen) {
            return await this._channelSession.sendChat(chat, noSeen);
        }
        forwardChat(chat, noSeen) {
            return this._channelSession.forwardChat(chat, noSeen);
        }
        deleteChat(chat) {
            return this._channelSession.deleteChat(chat);
        }
        async inviteUsers() {
            // Cannot invite users to open channel
            return { success: false, status: request_1.KnownDataStatusCode.OPERATION_DENIED };
        }
        syncChatList(endLogId, startLogId) {
            return this._channelSession.syncChatList(endLogId, startLogId);
        }
        getChatListFrom(startLogId) {
            return this._channelSession.getChatListFrom(startLogId);
        }
        markRead(chat) {
            return this._openChannelSession.markRead(chat);
        }
        setMeta(type, meta) {
            return this._channelSession.setMeta(type, meta);
        }
        setTitleMeta(title) {
            return this.setMeta(meta_1.KnownChannelMetaType.TITLE, title);
        }
        setNoticeMeta(notice) {
            return this.setMeta(meta_1.KnownChannelMetaType.NOTICE, notice);
        }
        /**
         * Set privileged settings.
         * Need to be owner of the channel to set.
         *
         * @param {PrivilegeMetaContent} content
         * @return {AsyncCommandResult<SetChannelMeta>}
         */
        setPrivilegeMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.PRIVILEGE, util_1.JsonUtil.stringifyLoseless(content));
        }
        setProfileMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.PROFILE, util_1.JsonUtil.stringifyLoseless(content));
        }
        setTvMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.TV, util_1.JsonUtil.stringifyLoseless(content));
        }
        setTvLiveMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.TV_LIVE, util_1.JsonUtil.stringifyLoseless(content));
        }
        setLiveTalkInfoMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.LIVE_TALK_INFO, util_1.JsonUtil.stringifyLoseless(content));
        }
        setLiveTalkCountMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.LIVE_TALK_COUNT, util_1.JsonUtil.stringifyLoseless(content));
        }
        setGroupMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.GROUP, util_1.JsonUtil.stringifyLoseless(content));
        }
        /**
         * Set bot meta
         *
         * @param {BotMetaContent} content
         * @return {AsyncCommandResult<SetChannelMeta>}
         */
        setBotMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.BOT, util_1.JsonUtil.stringifyLoseless(content));
        }
        setPushAlert(flag) {
            return this._channelSession.setPushAlert(flag);
        }
        chatON() {
            return this._openChannelSession.chatON();
        }
        getLatestChannelInfo() {
            return this._openChannelSession.getLatestChannelInfo();
        }
        getLatestUserInfo(...users) {
            return this._openChannelSession.getLatestUserInfo(...users);
        }
        getAllLatestUserInfo() {
            return this._openChannelSession.getAllLatestUserInfo();
        }
        getLatestOpenLink() {
            return this._openChannelSession.getLatestOpenLink();
        }
        createEvent(chat, type, count) {
            return this._openChannelSession.createEvent(chat, type, count);
        }
        getKickList() {
            return this._openChannelSession.getKickList();
        }
        removeKicked(user) {
            return this._openChannelSession.removeKicked(user);
        }
        setUserPerm(user, perm) {
            return this._openChannelSession.setUserPerm(user, perm);
        }
        handoverHost(user) {
            return this._openChannelSession.handoverHost(user);
        }
        kickUser(user) {
            return this._openChannelSession.kickUser(user);
        }
        blockUser(user) {
            return this._openChannelSession.blockUser(user);
        }
        react(flag) {
            return this._openChannelSession.react(flag);
        }
        getReaction() {
            return this._openChannelSession.getReaction();
        }
        changeProfile(profile) {
            return this._openChannelSession.changeProfile(profile);
        }
        hideChat(chat) {
            return this._openChannelSession.hideChat(chat);
        }
        downloadMedia(media, type, offset) {
            return this._channelSession.downloadMedia(media, type, offset);
        }
        downloadMediaThumb(media, type, offset) {
            return this._channelSession.downloadMediaThumb(media, type, offset);
        }
        uploadMedia(type, form) {
            return this._channelSession.uploadMedia(type, form);
        }
        uploadMultiMedia(type, forms) {
            return this._channelSession.uploadMultiMedia(type, forms);
        }
        sendMedia(type, template) {
            return channel_1.sendMedia(this._channelSession, type, template);
        }
        sendMultiMedia(type, templates) {
            return channel_1.sendMultiMedia(this._channelSession, type, templates);
        }
        async updateAll() {
            const infoRes = await this.getLatestChannelInfo();
            if (!infoRes.success)
                return infoRes;
            const linkRes = await this.getLatestOpenLink();
            if (!linkRes.success)
                return linkRes;
            return this.chatON();
        }
        // Called when broadcast packets are recevied.
        async pushReceived(method, data, parentCtx) {
            await this._handler.pushReceived(method, data, parentCtx);
            await this._openHandler.pushReceived(method, data, parentCtx);
        }
    }
    exports.TalkOpenChannel = TalkOpenChannel;
});
