// node-kakao/dist/talk/openlink/talk-open-channel-session.js
/*
 * Created on Wed Jan 27 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../openlink", "../../request", "../../packet/struct", "./talk-open-link-session", "../channel"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkOpenChannelManageSession = exports.TalkOpenChannelSession = void 0;
    const openlink_1 = require("../../openlink");
    const request_1 = require("../../request");
    const struct_1 = require("../../packet/struct");
    const talk_open_link_session_1 = require("./talk-open-link-session");
    const channel_1 = require("../channel");
    /**
     * Default OpenChannelSession implementation.
     */
    class TalkOpenChannelSession {
        constructor(channel, session) {
            this._channel = channel;
            this._session = session;
            this._normalSession = new channel_1.TalkNormalChannelSession(channel, session);
            this._linkSession = new talk_open_link_session_1.TalkOpenLinkSession(session);
        }
        get session() {
            return this._session;
        }
        chatON() {
            return this._normalSession.chatON();
        }
        async markRead(chat) {
            const status = (await this._session.request('NOTIREAD', {
                'chatId': this._channel.channelId,
                'li': this._channel.linkId,
                'watermark': chat.logId,
            })).status;
            return {
                success: status === request_1.KnownDataStatusCode.SUCCESS,
                status,
            };
        }
        async getLatestChannelInfo() {
            const res = await this._session.request('CHATINFO', {
                'chatId': this._channel.channelId,
            });
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS) {
                return { success: false, status: res.status };
            }
            return {
                success: true,
                status: res.status,
                result: struct_1.structToOpenChannelInfo(res.chatInfo),
            };
        }
        async getLatestUserInfo(...channelUsers) {
            const res = await this._session.request('MEMBER', {
                'chatId': this._channel.channelId,
                'memberIds': channelUsers.map((user) => user.userId),
            });
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS) {
                return { success: false, status: res.status };
            }
            const result = res.members.map((member) => struct_1.structToOpenChannelUserInfo(member));
            return { status: res.status, success: true, result };
        }
        async getAllLatestUserInfo() {
            const res = await this._session.request('GETMEM', {
                'chatId': this._channel.channelId,
            });
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS) {
                return { success: false, status: res.status };
            }
            const result = res.members.map((member) => struct_1.structToOpenChannelUserInfo(member));
            return { status: res.status, success: true, result };
        }
        getKickList() {
            return this._linkSession.getKickList(this._channel);
        }
        removeKicked(user) {
            return this._linkSession.removeKicked(this._channel, { ...user, kickedChannelId: this._channel.channelId });
        }
        react(flag) {
            return this._linkSession.react(this._channel, flag);
        }
        getReaction() {
            return this._linkSession.getReaction(this._channel);
        }
        async getLatestOpenLink() {
            const res = await this._linkSession.getOpenLink(this._channel);
            if (res.success) {
                return { success: true, status: res.status, result: res.result[0] };
            }
            else {
                return res;
            }
        }
        async setUserPerm(user, perm) {
            const res = await this._session.request('SETMEMTYPE', {
                'c': this._channel.channelId,
                'li': this._channel.linkId,
                'mids': [user.userId],
                'mts': [perm],
            });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async createEvent(chat, type, count) {
            const res = await this._session.request('RELAYEVENT', {
                'c': this._channel.channelId,
                'li': this._channel.linkId,
                'et': type,
                'ec': count,
                'logId': chat.logId,
                't': chat.type,
            });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async handoverHost(user) {
            const res = await this._session.request('SETMEMTYPE', {
                'c': this._channel.channelId,
                'li': this._channel.linkId,
                'mids': [user.userId, this._session.clientUser.userId],
                'mts': [openlink_1.OpenChannelUserPerm.OWNER, openlink_1.OpenChannelUserPerm.NONE],
            });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async kickUser(user) {
            const res = await this._session.request('KICKMEM', {
                'c': this._channel.channelId,
                'li': this._channel.linkId,
                'mid': user.userId,
            });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async blockUser(user) {
            const res = await this._session.request('BLIND', {
                'c': this._channel.channelId,
                'li': this._channel.linkId,
                'mid': user.userId,
                // Reporting user is not supported.
                'r': false,
            });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async changeProfile(profile) {
            const res = await this._session.request('UPLINKPROF', {
                'li': this._channel.linkId,
                ...openlink_1.OpenLinkProfile.serializeLinkProfile(profile),
            });
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { status: res.status, success: false };
            if (res['olu']) {
                return {
                    status: res.status,
                    success: true,
                    result: struct_1.structToOpenLinkChannelUserInfo(res['olu']),
                };
            }
            return { status: res.status, success: true, result: null };
        }
        async hideChat(chat) {
            const res = await this._session.request('REWRITE', {
                'li': this._channel.linkId,
                'c': this._channel.channelId,
                'logId': chat.logId,
                't': chat.type,
            });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
    }
    exports.TalkOpenChannelSession = TalkOpenChannelSession;
    class TalkOpenChannelManageSession {
        constructor(_session) {
            this._session = _session;
            this._normalSession = new channel_1.TalkChannelManageSession(_session);
        }
        leaveChannel(channel) {
            return this._normalSession.leaveChannel(channel);
        }
        async leaveKicked(channel) {
            const res = await this._session.request('KICKLEAVE', {
                'c': channel.channelId,
                'li': channel.linkId,
            });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        async joinChannel(link, profile, passcode) {
            let token;
            if (passcode) {
                const tokenRes = await this._session.request('CHECKJOIN', {
                    'li': link.linkId,
                    'pc': passcode,
                });
                if (tokenRes.status !== request_1.KnownDataStatusCode.SUCCESS)
                    return { status: tokenRes.status, success: false };
                token = tokenRes['tk'];
            }
            const reqData = {
                'li': link.linkId,
                'ref': 'EW:',
                ...openlink_1.OpenLinkProfile.serializeLinkProfile(profile),
            };
            if (token)
                reqData['tk'] = token;
            const res = await this._session.request('JOINLINK', reqData);
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { status: res.status, success: false };
            return { status: res.status, success: true, result: { channelId: res.chatRoom.chatId, linkId: res.ol.li } };
        }
    }
    exports.TalkOpenChannelManageSession = TalkOpenChannelManageSession;
});
