// node-kakao/dist/talk/openlink/talk-open-channel-list.js
/*
 * Created on Wed Jan 27 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "bson", "../../event", "../../request", "../channel", "./talk-open-channel", "./talk-open-channel-handler", "./talk-open-channel-session", "./client-link-store"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkOpenChannelList = void 0;
    const bson_1 = require("bson");
    const event_1 = require("../../event");
    const request_1 = require("../../request");
    const channel_1 = require("../channel");
    const talk_open_channel_1 = require("./talk-open-channel");
    const talk_open_channel_handler_1 = require("./talk-open-channel-handler");
    const talk_open_channel_session_1 = require("./talk-open-channel-session");
    const client_link_store_1 = require("./client-link-store");
    /**
     * Manage open profile, channel.
     */
    class TalkOpenChannelList extends event_1.TypedEmitter {
        constructor(_session, _loader, list, clientLinkList) {
            super();
            this._session = _session;
            this._loader = _loader;
            this._manageSession = new talk_open_channel_session_1.TalkOpenChannelManageSession(_session);
            this._handler = new channel_1.TalkChannelListHandler(this, this, this);
            this._openHandler = new talk_open_channel_handler_1.TalkOpenChannelListHandler(this, this, this);
            this._linkStore = new client_link_store_1.TalkClientLinkStore(_session, clientLinkList);
            this._map = new Map();
            if (list.length > 0) {
                list.forEach((channel) => this._map.set(channel.channelId.toString(), channel));
            }
        }
        get linkService() {
            return this._linkStore;
        }
        /**
         * @param {InformedOpenLink} link
         * @deprecated
         */
        addClientLink(link) {
            this._linkStore.addClientLink(link);
        }
        /**
         * @param {Long} linkId
         * @return {boolean}
         * @deprecated
         */
        deleteClientLink(linkId) {
            return this._linkStore.deleteClientLink(linkId);
        }
        removeChannel(channel) {
            return this._map.delete(channel.channelId.toString());
        }
        async addChannel(channel) {
            return this.addOpenChannel({ ...channel, linkId: bson_1.Long.ZERO });
        }
        get(channelId) {
            return this._map.get(channelId.toString());
        }
        /**
         * Find open channel using linkId
         *
         * @deprecated
         * @param {Long} linkId
         * @return {TalkOpenChannel | undefined}
         */
        getChannelByLinkId(linkId) {
            for (const channel of this.all()) {
                if (channel.linkId.eq(linkId))
                    return channel;
            }
        }
        /**
         * Find all open channel using same linkId
         *
         * @param {Long} linkId
         * @return {TalkOpenChannel[]}
         */
        getLinkChannelList(linkId) {
            const list = [];
            for (const channel of this.all()) {
                if (channel.linkId.eq(linkId))
                    list.push(channel);
            }
            return list;
        }
        get size() {
            return this._map.size;
        }
        all() {
            return this._map.values();
        }
        /**
         * @return {IterableIterator<InformedOpenLink>}
         * @deprecated
         */
        allClientLink() {
            return this._linkStore.allClientLink();
        }
        /**
         * @param {Long} linkId
         * @return {InformedOpenLink | undefined}
         * @deprecated
         */
        getClientLink(linkId) {
            return this._linkStore.getClientLink(linkId);
        }
        /**
         * @deprecated
         */
        get clientLinkCount() {
            return this._linkStore.clientLinkCount;
        }
        async addOpenChannel(channel, lastUpdate) {
            const last = this.get(channel.channelId);
            if (last)
                return { success: true, status: request_1.KnownDataStatusCode.SUCCESS, result: last };
            const infoStoreRes = await this._loader.loadOpenChannelStore(channel, lastUpdate);
            const chatStoreRes = await this._loader.loadChatListStore(channel);
            const talkChannel = new talk_open_channel_1.TalkOpenChannel(channel, this._session, infoStoreRes.value, chatStoreRes.value);
            this._map.set(channel.channelId.toString(), talkChannel);
            if (infoStoreRes.shouldUpdate) {
                const res = await talkChannel.updateAll();
                if (!res.success)
                    return res;
            }
            if (chatStoreRes.shouldUpdate) {
                await channel_1.updateChatList(talkChannel);
            }
            return { success: true, status: request_1.KnownDataStatusCode.SUCCESS, result: talkChannel };
        }
        async leaveKicked(channel) {
            const res = await this._manageSession.leaveKicked(channel);
            if (res.success) {
                this._map.delete(channel.channelId.toString());
            }
            return res;
        }
        async leaveChannel(channel) {
            const res = await this._manageSession.leaveChannel(channel);
            if (res.success) {
                this._map.delete(channel.channelId.toString());
            }
            return res;
        }
        getLatestLinkList() {
            return this._linkStore.getLatestLinkList();
        }
        getOpenLink(...components) {
            return this._linkStore.getOpenLink(...components);
        }
        getJoinInfo(linkURL, referer) {
            return this._linkStore.getJoinInfo(linkURL, referer);
        }
        getKickList(link) {
            return this._linkStore.getKickList(link);
        }
        removeKicked(link, kickedUser) {
            return this._linkStore.removeKicked(link, kickedUser);
        }
        async deleteLink(link) {
            const res = await this._linkStore.deleteLink(link);
            if (res.success) {
                this._linkStore.deleteClientLink(link.linkId);
                for (const channel of this.getLinkChannelList(link.linkId)) {
                    this.removeChannel(channel);
                }
            }
            return res;
        }
        react(link, flag) {
            return this._linkStore.react(link, flag);
        }
        getReaction(link) {
            return this._linkStore.getReaction(link);
        }
        async createOpenChannel(template, profile) {
            const res = await this._linkStore.createOpenChannel(template, profile);
            if (!res.success)
                return res;
            return this.addOpenChannel(res.result);
        }
        createOpenDirectProfile(template, profile) {
            return this._linkStore.createOpenDirectProfile(template, profile);
        }
        createOpenProfile(template) {
            return this._linkStore.createOpenProfile(template);
        }
        updateOpenLink(link, settings) {
            return this._linkStore.updateOpenLink(link, settings);
        }
        async pushReceived(method, data, parentCtx) {
            const ctx = new event_1.EventContext(this, parentCtx);
            await Promise.all(Array.from(this._map.values()).map((channel) => channel.pushReceived(method, data, ctx)));
            await this._handler.pushReceived(method, data, parentCtx);
            await this._openHandler.pushReceived(method, data, parentCtx);
            await this._linkStore.pushReceived(method, data, parentCtx);
        }
        async joinChannel(link, profile, passcode) {
            const res = await this._manageSession.joinChannel(link, profile, passcode);
            if (!res.success)
                return res;
            return this.addOpenChannel(res.result);
        }
        /**
         * Initialize TalkChannelList using channelList.
         * @param {TalkOpenChannelList} talkChannelList
         * @param {LoginData<OpenChannelData>[]} channelList
         */
        static async initialize(talkChannelList, channelList = []) {
            talkChannelList._map.clear();
            await Promise.all(channelList.map((data) => talkChannelList.addOpenChannel(data.channel, data.lastUpdate)));
            await talkChannelList._linkStore.getLatestLinkList();
            return talkChannelList;
        }
    }
    exports.TalkOpenChannelList = TalkOpenChannelList;
});
