// node-kakao/dist/talk/client/talk-client-session.js
/*
 * Created on Fri Jan 22 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../request", "../../packet/struct", "bson"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkClientSession = void 0;
    const request_1 = require("../../request");
    const struct_1 = require("../../packet/struct");
    const bson_1 = require("bson");
    class TalkClientSession {
        constructor(_session, configuration) {
            this._session = _session;
            this.configuration = configuration;
            this._lastLoginRev = 0;
        }
        get session() {
            return this._session;
        }
        async login(credential) {
            const config = this.configuration;
            const req = {
                'appVer': config.appVersion,
                'prtVer': '1',
                'os': config.agent,
                'lang': config.language,
                'duuid': credential.deviceUUID,
                'oauthToken': credential.accessToken,
                'dtype': config.deviceType,
                'ntype': config.netType,
                'MCCMNC': config.mccmnc,
                'revision': this._lastLoginRev,
                'chatIds': [],
                'maxIds': [],
                'lastTokenId': bson_1.Long.ZERO,
                'lbk': 0,
                'rp': null,
                'bg': false,
            };
            const loginRes = await this._session.request('LOGINLIST', req);
            if (loginRes.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { status: loginRes.status, success: false };
            let status = loginRes.status;
            const chatDataList = loginRes.chatDatas;
            const delChannelIdList = loginRes.delChatIds;
            this._lastLoginRev = loginRes.revision;
            let lastRes = loginRes;
            while (!lastRes.eof) {
                const res = await this._session.request('LCHATLIST', {
                    'lastTokenId': lastRes.lastTokenId,
                    'lastChatId': lastRes.lastChatId
                });
                if (loginRes.status !== request_1.KnownDataStatusCode.SUCCESS) {
                    status = request_1.KnownDataStatusCode.LOGINLIST_CHATLIST_FAILED;
                    break;
                }
                chatDataList.push(...res.chatDatas);
                delChannelIdList.push(...res.delChatIds);
                lastRes = res;
            }
            const channelList = [];
            for (const channelData of chatDataList) {
                let channel;
                if (channelData.li) {
                    channel = {
                        channelId: channelData.c,
                        linkId: channelData.li,
                        info: struct_1.dataStructToOpenChannelInfo(channelData)
                    };
                }
                else {
                    channel = {
                        channelId: channelData.c,
                        info: struct_1.dataStructToNormalChannelInfo(channelData)
                    };
                }
                channelList.push({
                    lastUpdate: channelData.o,
                    channel
                });
            }
            return {
                status,
                success: true,
                result: {
                    channelList,
                    lastChannelId: lastRes.lastChatId,
                    lastTokenId: lastRes.lastTokenId,
                    mcmRevision: lastRes.mcmRevision,
                    revision: loginRes.revision,
                    revisionInfo: loginRes.revisionInfo,
                    removedChannelIdList: delChannelIdList,
                    minLogId: loginRes.minLogId,
                    userId: loginRes.userId,
                },
            };
        }
        async setStatus(status) {
            const res = await this._session.request('SETST', { st: status });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS };
        }
        /**
         * Unknown
         *
         * @param {number[]} unknown
         * @return {AsyncCommandResult<DefaultRes>}
         */
        async getTokens(unknown) {
            const res = await this._session.request('GETTOKEN', { ts: unknown });
            return { status: res.status, success: res.status === request_1.KnownDataStatusCode.SUCCESS, result: res };
        }
    }
    exports.TalkClientSession = TalkClientSession;
});
