// node-kakao/dist/talk/channel/talk-normal-channel.js
/*
 * Created on Mon Jan 18 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./talk-normal-channel-session", "../../event", "./talk-channel-handler", "../../channel/meta", "../../util", "./talk-normal-channel-data-session", "./talk-channel-data-session", "./talk-channel-session", "./talk-normal-channel-handler", "./common"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkNormalChannel = void 0;
    const talk_normal_channel_session_1 = require("./talk-normal-channel-session");
    const event_1 = require("../../event");
    const talk_channel_handler_1 = require("./talk-channel-handler");
    const meta_1 = require("../../channel/meta");
    const util_1 = require("../../util");
    const talk_normal_channel_data_session_1 = require("./talk-normal-channel-data-session");
    const talk_channel_data_session_1 = require("./talk-channel-data-session");
    const talk_channel_session_1 = require("./talk-channel-session");
    const talk_normal_channel_handler_1 = require("./talk-normal-channel-handler");
    const common_1 = require("./common");
    class TalkNormalChannel extends event_1.TypedEmitter {
        constructor(_channel, session, store, _chatListStore) {
            super();
            this._channel = _channel;
            this._chatListStore = _chatListStore;
            this._channelSession = new talk_channel_data_session_1.TalkChannelDataSession(session.clientUser, new talk_channel_session_1.TalkChannelSession(this, session), store, _chatListStore);
            const normalSession = new talk_normal_channel_session_1.TalkNormalChannelSession(this, session);
            this._normalChannelSession = new talk_normal_channel_data_session_1.TalkNormalChannelDataSession(session.clientUser, normalSession, store);
            this._handler = new talk_channel_handler_1.TalkChannelHandler(this, this, store, _chatListStore);
            this._normalHandler = new talk_normal_channel_handler_1.TalkNormalChannelHandler(this, normalSession, this, store, _chatListStore);
        }
        get clientUser() {
            return this._normalChannelSession.clientUser;
        }
        get channelId() {
            return this._channel.channelId;
        }
        get chatListStore() {
            return this._chatListStore;
        }
        get store() {
            return this._normalChannelSession.store;
        }
        get info() {
            return this.store.info;
        }
        get userCount() {
            return this.store.userCount;
        }
        getName() {
            const nameMeta = this.info.metaMap[meta_1.KnownChannelMetaType.TITLE];
            return nameMeta && nameMeta.content || '';
        }
        getDisplayName() {
            return this.getName() || this.info.displayUserList.map((user) => user.nickname).join(', ');
        }
        getUserInfo(user) {
            return this._normalChannelSession.store.getUserInfo(user);
        }
        getAllUserInfo() {
            return this._normalChannelSession.store.getAllUserInfo();
        }
        getReadCount(chat) {
            return this.store.getReadCount(chat);
        }
        getReaders(chat) {
            return this.store.getReaders(chat);
        }
        sendChat(chat, noSeen) {
            return this._channelSession.sendChat(chat, noSeen);
        }
        forwardChat(chat, noSeen) {
            return this._channelSession.forwardChat(chat, noSeen);
        }
        deleteChat(chat) {
            return this._channelSession.deleteChat(chat);
        }
        markRead(chat) {
            return this._channelSession.markRead(chat);
        }
        setMeta(type, meta) {
            return this._channelSession.setMeta(type, meta);
        }
        setTitleMeta(title) {
            return this.setMeta(meta_1.KnownChannelMetaType.TITLE, title);
        }
        setNoticeMeta(notice) {
            return this.setMeta(meta_1.KnownChannelMetaType.NOTICE, notice);
        }
        setProfileMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.PROFILE, util_1.JsonUtil.stringifyLoseless(content));
        }
        setTvMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.TV, util_1.JsonUtil.stringifyLoseless(content));
        }
        setTvLiveMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.TV_LIVE, util_1.JsonUtil.stringifyLoseless(content));
        }
        setLiveTalkInfoMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.LIVE_TALK_INFO, util_1.JsonUtil.stringifyLoseless(content));
        }
        setLiveTalkCountMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.LIVE_TALK_COUNT, util_1.JsonUtil.stringifyLoseless(content));
        }
        setGroupMeta(content) {
            return this.setMeta(meta_1.KnownChannelMetaType.GROUP, util_1.JsonUtil.stringifyLoseless(content));
        }
        setPushAlert(flag) {
            return this._channelSession.setPushAlert(flag);
        }
        inviteUsers(users) {
            return this._normalChannelSession.inviteUsers(users);
        }
        syncChatList(endLogId, startLogId) {
            return this._channelSession.syncChatList(endLogId, startLogId);
        }
        getChatListFrom(startLogId) {
            return this._channelSession.getChatListFrom(startLogId);
        }
        async chatON() {
            return this._normalChannelSession.chatON();
        }
        async getLatestChannelInfo() {
            return this._normalChannelSession.getLatestChannelInfo();
        }
        async getLatestUserInfo(...users) {
            return this._normalChannelSession.getLatestUserInfo(...users);
        }
        async getAllLatestUserInfo() {
            return this._normalChannelSession.getAllLatestUserInfo();
        }
        downloadMedia(media, type, offset) {
            return this._channelSession.downloadMedia(media, type, offset);
        }
        downloadMediaThumb(media, type, offset) {
            return this._channelSession.downloadMediaThumb(media, type, offset);
        }
        uploadMedia(type, form) {
            return this._channelSession.uploadMedia(type, form);
        }
        uploadMultiMedia(type, forms) {
            return this._channelSession.uploadMultiMedia(type, forms);
        }
        sendMedia(type, template) {
            return common_1.sendMedia(this._channelSession, type, template);
        }
        sendMultiMedia(type, templates) {
            return common_1.sendMultiMedia(this._channelSession, type, templates);
        }
        async updateAll() {
            const infoRes = await this.getLatestChannelInfo();
            if (!infoRes.success)
                return infoRes;
            return this.chatON();
        }
        async pushReceived(method, data, parentCtx) {
            await this._handler.pushReceived(method, data, parentCtx);
            await this._normalHandler.pushReceived(method, data, parentCtx);
        }
    }
    exports.TalkNormalChannel = TalkNormalChannel;
});
