// node-kakao/dist/talk/channel/talk-normal-channel-data-session.js
/*
 * Created on Sat Mar 06 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../channel", "../../packet/struct", "./common"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkNormalChannelDataSession = void 0;
    const channel_1 = require("../../channel");
    const struct_1 = require("../../packet/struct");
    const common_1 = require("./common");
    /**
     * Do normal channel session operations and updates store.
     */
    class TalkNormalChannelDataSession {
        constructor(_clientUser, _channelSession, _store) {
            this._clientUser = _clientUser;
            this._channelSession = _channelSession;
            this._store = _store;
        }
        get clientUser() {
            return this._clientUser;
        }
        get store() {
            return this._store;
        }
        async inviteUsers(users) {
            const res = await this._channelSession.inviteUsers(users);
            if (res.success) {
                await this.getLatestUserInfo(...users);
            }
            return res;
        }
        async chatON() {
            const res = await this._channelSession.chatON();
            if (res.success) {
                const { result } = res;
                if (this._store.info.type !== result.t || this._store.info.lastChatLogId !== result.l) {
                    this._store.updateInfo({ type: result.t, lastChatLogId: result.l });
                }
                if (result.a && result.w) {
                    this._store.clearWatermark();
                    common_1.initWatermark(this._store, result.a, result.w);
                }
                if (result.m) {
                    this._store.clearUserList();
                    const structList = result.m;
                    for (const struct of structList) {
                        const wrapped = struct_1.structToChannelUserInfo(struct);
                        this._store.updateUserInfo(wrapped, wrapped);
                    }
                }
                else if (result.mi) {
                    this._store.clearUserList();
                    const userInitres = await common_1.initNormalUserList(this._channelSession, result.mi);
                    if (!userInitres.success)
                        return userInitres;
                    for (const info of userInitres.result) {
                        this._store.updateUserInfo(info, info);
                    }
                    const channelSession = this._channelSession;
                    if (!this._store.getUserInfo(this._clientUser)) {
                        const clientRes = await channelSession.getLatestUserInfo(this._clientUser);
                        if (!clientRes.success)
                            return clientRes;
                        for (const user of clientRes.result) {
                            this._store.updateUserInfo(user, user);
                        }
                    }
                }
            }
            return res;
        }
        async getLatestChannelInfo() {
            const infoRes = await this._channelSession.getLatestChannelInfo();
            if (infoRes.success) {
                this._store.setInfo(channel_1.NormalChannelInfo.createPartial(infoRes.result));
            }
            return infoRes;
        }
        async getLatestUserInfo(...users) {
            const infoRes = await this._channelSession.getLatestUserInfo(...users);
            if (infoRes.success) {
                const result = infoRes.result;
                result.forEach((info) => this._store.updateUserInfo(info, info));
            }
            return infoRes;
        }
        async getAllLatestUserInfo() {
            const infoRes = await this._channelSession.getAllLatestUserInfo();
            if (infoRes.success) {
                infoRes.result.map((info) => this._store.updateUserInfo(info, info));
            }
            return infoRes;
        }
    }
    exports.TalkNormalChannelDataSession = TalkNormalChannelDataSession;
});
