// node-kakao/dist/talk/channel/talk-normal-channel-session.js
/*
 * Created on Fri Jan 22 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../request", "../../packet/struct", "bson"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkNormalChannelSession = void 0;
    const request_1 = require("../../request");
    const struct_1 = require("../../packet/struct");
    const bson_1 = require("bson");
    /**
     * Default NormalChannelSession implementation
     */
    class TalkNormalChannelSession {
        constructor(channel, session) {
            this._channel = channel;
            this._session = session;
            this.currentMsgId = 0;
        }
        get session() {
            return this._session;
        }
        async chatON() {
            const res = await this._session.request('CHATONROOM', {
                'chatId': this._channel.channelId,
                'token': bson_1.Long.ZERO,
                'opt': bson_1.Long.ZERO,
            });
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { success: false, status: res.status };
            return { success: true, status: res.status, result: res };
        }
        async inviteUsers(users) {
            const { status } = await this._session.request('ADDMEM', {
                'chatId': this._channel.channelId,
                'memberIds': users.map((user) => user.userId),
            });
            return {
                success: status === request_1.KnownDataStatusCode.SUCCESS,
                status,
            };
        }
        async getLatestChannelInfo() {
            const res = await this._session.request('CHATINFO', {
                'chatId': this._channel.channelId,
            });
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { success: false, status: res.status };
            return {
                success: true,
                status: res.status,
                result: struct_1.structToNormalChannelInfo(res.chatInfo),
            };
        }
        async getLatestUserInfo(...channelUsers) {
            const res = await this._session.request('MEMBER', {
                'chatId': this._channel.channelId,
                'memberIds': channelUsers.map((user) => user.userId),
            });
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { success: false, status: res.status };
            const result = res.members.map((member) => struct_1.structToChannelUserInfo(member));
            return { success: true, status: res.status, result };
        }
        async getAllLatestUserInfo() {
            const res = await this._session.request('GETMEM', {
                'chatId': this._channel.channelId,
            });
            if (res.status !== request_1.KnownDataStatusCode.SUCCESS)
                return { success: false, status: res.status };
            const result = res.members.map((member) => struct_1.structToChannelUserInfo(member));
            return { success: true, status: res.status, result };
        }
    }
    exports.TalkNormalChannelSession = TalkNormalChannelSession;
});
