// node-kakao/dist/talk/channel/talk-normal-channel-list.js
/*
 * Created on Wed Jan 20 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../event", "../../request", "./talk-normal-channel", "./talk-channel-handler", "./talk-channel-session", "./common"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkNormalChannelList = void 0;
    const event_1 = require("../../event");
    const request_1 = require("../../request");
    const talk_normal_channel_1 = require("./talk-normal-channel");
    const talk_channel_handler_1 = require("./talk-channel-handler");
    const talk_channel_session_1 = require("./talk-channel-session");
    const common_1 = require("./common");
    /**
     * Manage session channels
     */
    class TalkNormalChannelList extends event_1.TypedEmitter {
        /**
         * Construct managed normal channel list
         * @param {TalkSession} _session
         * @param {ClientDataLoader} _loader
         * @param {TalkNormalChannel[]} list
         */
        constructor(_session, _loader, list) {
            super();
            this._session = _session;
            this._loader = _loader;
            this._handler = new talk_channel_handler_1.TalkChannelListHandler(this, this, this);
            this._manageSession = new talk_channel_session_1.TalkChannelManageSession(_session);
            this._map = new Map();
            if (list.length > 0) {
                list.forEach((channel) => this._map.set(channel.channelId.toString(), channel));
            }
        }
        get size() {
            return this._map.size;
        }
        get(channelId) {
            const strId = channelId.toString();
            return this._map.get(strId);
        }
        all() {
            return this._map.values();
        }
        async addChannel(channel, lastUpdate) {
            const last = this.get(channel.channelId);
            if (last)
                return { success: true, status: request_1.KnownDataStatusCode.SUCCESS, result: last };
            const strId = channel.channelId.toString();
            const infoStoreRes = await this._loader.loadNormalChannelStore(channel, lastUpdate);
            const chatStoreRes = await this._loader.loadChatListStore(channel);
            const talkChannel = new talk_normal_channel_1.TalkNormalChannel(channel, this._session, infoStoreRes.value, chatStoreRes.value);
            this._map.set(strId, talkChannel);
            if (infoStoreRes.shouldUpdate) {
                const res = await talkChannel.updateAll();
                if (!res.success)
                    return res;
            }
            if (chatStoreRes.shouldUpdate) {
                await common_1.updateChatList(talkChannel);
            }
            return { success: true, status: request_1.KnownDataStatusCode.SUCCESS, result: talkChannel };
        }
        removeChannel(channel) {
            return this._map.delete(channel.channelId.toString());
        }
        async createChannel(template) {
            const res = await this._manageSession.createChannel(template);
            if (!res.success)
                return res;
            return this.addChannel(res.result);
        }
        async createMemoChannel() {
            const res = await this._manageSession.createMemoChannel();
            if (!res.success)
                return res;
            return this.addChannel(res.result);
        }
        async leaveChannel(channel, block) {
            const res = await this._manageSession.leaveChannel(channel, block);
            if (res.success) {
                this.removeChannel(channel);
            }
            return res;
        }
        async pushReceived(method, data, parentCtx) {
            const ctx = new event_1.EventContext(this, parentCtx);
            await Promise.all(Array.from(this._map.values()).map((channel) => channel.pushReceived(method, data, ctx)));
            await this._handler.pushReceived(method, data, parentCtx);
        }
        /**
         * Initialize TalkChannelList using channelList.
         * @param {TalkNormalChannelList} talkChannelList
         * @param {LoginData<NormalChannelData>[]} channelList
         */
        static async initialize(talkChannelList, channelList = []) {
            talkChannelList._map.clear();
            await Promise.all(channelList.map((data) => talkChannelList.addChannel(data.channel, data.lastUpdate)));
            return talkChannelList;
        }
    }
    exports.TalkNormalChannelList = TalkNormalChannelList;
});
