// node-kakao/dist/talk/channel/talk-channel-handler.js
/*
 * Created on Sat Jan 23 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../../chat", "../../packet/struct", "../chat"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkChannelListHandler = exports.TalkChannelHandler = void 0;
    const chat_1 = require("../../chat");
    const struct_1 = require("../../packet/struct");
    const chat_2 = require("../chat");
    /**
     * Capture and handle pushes coming to channel
     */
    class TalkChannelHandler {
        constructor(_channel, _emitter, _store, _chatListStore) {
            this._channel = _channel;
            this._emitter = _emitter;
            this._store = _store;
            this._chatListStore = _chatListStore;
        }
        get info() {
            return this._store.info;
        }
        _callEvent(parentCtx, event, ...args) {
            this._emitter.emit(event, ...args);
            parentCtx.emit(event, ...args);
        }
        _msgHandler(msgData, parentCtx) {
            if (!this._channel.channelId.equals(msgData.chatId))
                return;
            const chatLog = struct_1.structToChatlog(msgData.chatLog);
            if (msgData.authorNickname) {
                const userInfo = this._store.getUserInfo(chatLog.sender);
                if (userInfo && userInfo.nickname !== msgData.authorNickname) {
                    this._store.updateUserInfo(chatLog.sender, { nickname: msgData.authorNickname });
                }
            }
            this._callEvent(parentCtx, 'chat', new chat_2.TalkChatData(chatLog), this._channel);
            this._chatListStore.addChat(chatLog).then(() => {
                this._store.updateInfo({
                    lastChatLogId: msgData.logId,
                    lastChatLog: chatLog,
                });
            });
        }
        _feedHandler(data, parentCtx) {
            const channelId = data['c'];
            if (!this._channel.channelId.equals(channelId))
                return;
            const chatLog = struct_1.structToChatlog(data['chatLog']);
            this._callEvent(parentCtx, 'chat', new chat_2.TalkChatData(chatLog), this._channel);
            this._chatListStore.addChat(chatLog).then(() => {
                this._store.updateInfo({
                    lastChatLogId: chatLog.logId,
                    lastChatLog: chatLog,
                });
            });
        }
        _chatReadHandler(readData, parentCtx) {
            if (!this._channel.channelId.equals(readData.chatId))
                return;
            const reader = this._store.getUserInfo({ userId: readData.userId });
            this._store.updateWatermark(readData.userId, readData.watermark);
            this._callEvent(parentCtx, 'chat_read', { logId: readData.watermark }, this._channel, reader);
        }
        _metaChangeHandler(metaData, parentCtx) {
            if (!this._channel.channelId.equals(metaData.chatId))
                return;
            const metaType = metaData.meta.type;
            const meta = metaData.meta;
            this._callEvent(parentCtx, 'meta_change', this._channel, metaType, meta);
            const metaMap = { ...this.info.metaMap };
            metaMap[metaType] = meta;
            this._store.updateInfo({
                metaMap,
            });
        }
        _userLeftHandler(data, parentCtx) {
            const struct = data['chatLog'];
            if (!struct || !this._channel.channelId.eq(struct.chatId))
                return;
            const chatLog = struct_1.structToChatlog(struct);
            // TODO: The event should be called whatever the chat is valid or not.
            if (chatLog.type === chat_1.KnownChatType.FEED) {
                const feed = chat_1.feedFromChat(chatLog);
                if ('member' in feed) {
                    const memberFeed = feed;
                    const user = this._store.getUserInfo(memberFeed.member);
                    if (user) {
                        this._store.removeUser(user);
                        this._callEvent(parentCtx, 'user_left', chatLog, this._channel, user, feed);
                    }
                }
            }
            this._chatListStore.addChat(chatLog).then();
            return;
        }
        _msgDeleteHandler(data, parentCtx) {
            const struct = data['chatLog'];
            if (!this._channel.channelId.eq(struct.chatId))
                return;
            const chatLog = struct_1.structToChatlog(struct);
            if (chatLog.type !== chat_1.KnownChatType.FEED)
                return;
            const feed = chat_1.feedFromChat(chatLog);
            if (feed.feedType !== chat_1.KnownFeedType.DELETE_TO_ALL)
                return;
            const delAllFeed = feed;
            this._callEvent(parentCtx, 'chat_deleted', chatLog, this._channel, delAllFeed);
            this._chatListStore.addChat(chatLog).then(async () => {
                if (!delAllFeed.logId)
                    return;
                const chat = await this._chatListStore.get(delAllFeed.logId);
                if (!chat)
                    return;
                await this._chatListStore.updateChat(delAllFeed.logId, { type: chat.type | chat_1.DELETED_MESSAGE_OFFSET });
            }).then();
        }
        async pushReceived(method, data, parentCtx) {
            switch (method) {
                case 'MSG':
                    await this._msgHandler(data, parentCtx);
                    break;
                case 'FEED':
                    await this._feedHandler(data, parentCtx);
                    break;
                case 'DECUNREAD':
                    await this._chatReadHandler(data, parentCtx);
                    break;
                case 'CHGMETA':
                    await this._metaChangeHandler(data, parentCtx);
                    break;
                case 'DELMEM':
                    await this._userLeftHandler(data, parentCtx);
                    break;
                case 'SYNCDLMSG':
                    await this._msgDeleteHandler(data, parentCtx);
                    break;
            }
        }
    }
    exports.TalkChannelHandler = TalkChannelHandler;
    class TalkChannelListHandler {
        constructor(_list, _emitter, _updater) {
            this._list = _list;
            this._emitter = _emitter;
            this._updater = _updater;
        }
        _callEvent(parentCtx, event, ...args) {
            this._emitter.emit(event, ...args);
            parentCtx.emit(event, ...args);
        }
        async pushReceived(method, data, parentCtx) {
            switch (method) {
                case 'LEFT': {
                    const leftData = data;
                    const channel = this._list.get(leftData.chatId);
                    if (!channel)
                        return;
                    this._updater.removeChannel(channel);
                    this._callEvent(parentCtx, 'channel_left', channel);
                    break;
                }
                default: break;
            }
        }
    }
    exports.TalkChannelListHandler = TalkChannelListHandler;
});
